File: problem780.dart
import 'dart:math';

class PrimePower {
  final int p;
  final int r;
  PrimePower(this.p, this.r);
}

typedef VecInt = List<int>;
typedef VecLong = List<int>;
typedef IntPair = List<int>;
typedef PrimeFactors = List<int>;
typedef PrimeFactorisation = List<PrimePower>;

const int N = 1e9.toInt();
const int N_2 = N ~/ 2;
const int M = 1e9.toInt() + 7;

int SQ(int x) => (x * x) % M;

int INT(double x) => x.toInt();

int kMax = INT(N_2 / sqrt(3));

List<int> smallestPrimeFactor = List.filled(kMax + 1, 0);

PrimeFactorisation makePrimeFactorisation(int n) {
  final pf = <PrimePower>[];
  while (n > 1) {
    int p = smallestPrimeFactor[n];
    if (p == 0) {
      pf.add(PrimePower(n, 1));
      return pf;
    }
    int r = 0;
    while (n % p == 0) {
      n ~/= p;
      r++;
    }
    pf.add(PrimePower(p, r));
  }
  return pf;
}

PrimeFactors makePrimeFactors(PrimeFactorisation primeFactorisation) {
  return primeFactorisation.map((pp) => pp.p).toList();
}

int intPow(int x, int p) {
  if (p == 0) return 1;
  if (p == 1) return x;

  int tmp = intPow(x, p ~/ 2);
  if ((p % 2) == 0) {
    return tmp * tmp;
  } else {
    return x * tmp * tmp;
  }
}

void buildKIProd(int k, int l, PrimeFactorisation pf, int K_I, int prod) {
  int K_J = 1, L_J = 1, n = l;
  for (var pp in pf) {
    if (pp.r == 1) continue;
    int s = 0;
    while (n % pp.p == 0) {
      n ~/= pp.p;
      s++;
    }
    if (s < pp.r - 1) {
      K_J *= intPow(pp.p, pp.r);
      L_J *= intPow(pp.p, s);
      prod *= 2 * (s + 1);
    }
  }
  K_I = k * l ~/ (K_J * L_J);
}

int totient(int m, PrimeFactors primeFactors) {
  if (m == 1) return 1;
  int tot = m;
  for (int p in primeFactors) {
    tot ~/= p;
    tot *= (p - 1);
  }
  return tot;
}

int phi(int m, int n, PrimeFactors primeFactors) {
  if (m == 1) {
    return n;
  } else if (n <= 1) {
    return n;
  } else if (n % m == 0) {
    return (n ~/ m) * totient(m, primeFactors);
  }
  int res = n;
  for (int i = 1; i < (1 << primeFactors.length); ++i) {
    int k = i.bitCount;
    int denom = 1;
    for (int j = 0; (1 << j) <= i; ++j) {
      if ((i & (1 << j)) != 0) {
        denom *= primeFactors[j];
      }
    }
    res += (1 - 2 * (k % 2)) * (n ~/ denom);
  }
  return res;
}

int sumOfDiv(int n) {
  int res = 0;
  int sqrtNFloor = sqrt(n).toInt();
  for (int m = 1; m <= sqrtNFloor; ++m) {
    res += n ~/ m;
    res %= M;
  }
  res *= 2;
  res %= M;
  res -= sqrtNFloor * sqrtNFloor;
  res %= M;
  if (res < 0) {
    res += M;
  }
  return res;
}

void main() {
  print("Building smallest prime factor sieve");

  smallestPrimeFactor[0] = 1;
  smallestPrimeFactor[1] = 1;
  for (int p = 2; p <= kMax; ++p) {
    if (smallestPrimeFactor[p] != 0) continue;
    for (int m = 2; m * p <= kMax; ++m) {
      smallestPrimeFactor[m * p] = p;
    }
  }

  print("Building number of divisors");
  List<int> ndiv = List.filled(kMax + 1, 2);
  ndiv[0] = 0;
  ndiv[1] = 1;
  for (int n = 2; n <= sqrt(kMax).toInt(); ++n) {
    ndiv[n * n]++;
    for (int m = n + 1; m <= kMax ~/ n; ++m) {
      ndiv[m * n] += 2;
    }
  }

  int res1 = sumOfDiv(N_2);

  print("Result: $res1");
}

File: problem_780.py
import math
from collections import namedtuple

PrimePower = namedtuple('PrimePower', ['p', 'r'])

N = int(1e9)
N_2 = N // 2
M = int(1e9) + 7

kMax = int(N_2 / math.sqrt(3))

smallest_prime_factor = [0] * (kMax + 1)


def sq(x):
    return (x * x) % M


def int_pow(x, p):
    if p == 0:
        return 1
    if p == 1:
        return x

    tmp = int_pow(x, p // 2)
    if (p % 2) == 0:
        return tmp * tmp
    else:
        return x * tmp * tmp


# Generate prime factorization
def make_prime_factorisation(n):
    pf = []
    while n > 1:
        p = smallest_prime_factor[n]
        if p == 0:
            pf.append(PrimePower(n, 1))
            return pf
        r = 0
        while n % p == 0:
            n //= p
            r += 1
        pf.append(PrimePower(p, r))
    return pf


def make_prime_factors(prime_factorisation):
    return [pp.p for pp in prime_factorisation]


def build_ki_prod(k, l, pf, K_I, prod):
    K_J = 1
    L_J = 1
    n = l
    for pp in pf:
        if pp.r == 1:
            continue
        s = 0
        while n % pp.p == 0:
            n //= pp.p
            s += 1
        if s < pp.r - 1:
            K_J *= int_pow(pp.p, pp.r)
            L_J *= int_pow(pp.p, s)
            prod *= 2 * (s + 1)
    K_I = k * l // (K_J * L_J)


def totient(m, prime_factors):
    if m == 1:
        return 1
    tot = m
    for p in prime_factors:
        tot //= p
        tot *= (p - 1)
    return tot


def phi(m, n, prime_factors):
    if m == 1:
        return n
    elif n <= 1:
        return n
    elif n % m == 0:
        return (n // m) * totient(m, prime_factors)
    res = n
    for i in range(1, 1 << len(prime_factors)):
        k = bin(i).count('1')
        denom = 1
        for j in range(len(prime_factors)):
            if (i & (1 << j)) != 0:
                denom *= prime_factors[j]
        res += (1 - 2 * (k % 2)) * (n // denom)
    return res


def sum_of_div(n):
    res = 0
    sqrt_n_floor = int(math.sqrt(n))
    for m in range(1, sqrt_n_floor + 1):
        res += n // m
        res %= M
    res *= 2
    res %= M
    res -= sqrt_n_floor * sqrt_n_floor
    res %= M
    if res < 0:
        res += M
    return res


if __name__ == "__main__":
    print("Building smallest prime factor sieve")

    smallest_prime_factor[0] = 1
    smallest_prime_factor[1] = 1
    for p in range(2, kMax + 1):
        if smallest_prime_factor[p] != 0:
            continue
        for m in range(2, (kMax // p) + 1):
            smallest_prime_factor[m * p] = p

    print("Building number of divisors")
    ndiv = [2] * (kMax + 1)
    ndiv[0] = 0
    ndiv[1] = 1
    for n in range(2, int(math.sqrt(kMax)) + 1):
        ndiv[n * n] += 1
        for m in range(n + 1, (kMax // n) + 1):
            ndiv[m * n] += 2

    res1 = sum_of_div(N_2)

    print("Result:", res1)